### DISCUSSION TITLE
How and where exactly can be used the Kit related variables?

### DISCUSSION THREAD

**Discussion by HofiOne** [microsoft/vscode-cmake-tools#3379] [Q&A] [answered]
Hi,

I'm trying to use [CMake kits](https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/kits.md#cmake-kits) related variables mentioned in [Configure CMake Tools settings](https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/cmake-settings.md#configure-cmake-tools-settings) in my tasks.json and settings.json, but seems they are not really defined (yet?) when referencing them.

As I understood, the simplest test case should work also, e.g. when I echo the content of the variables from a simple task
```
{
	"label": "echo",
	"type": "shell",
	"command": "echo ${buildKit}, ${generator}, ${workspaceRoot}"
},
```
but using the above test task.json when invoking the `echo` task it prints only the value of the ${workspaceRoot} variable (that I guess comes from the globally defined VSCode variable, not from the CMake Tools defined set)

I also tried it in my `settings.json`
[Configure CMake Tools settings](https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/cmake-settings.md#configure-cmake-tools-settings) mentions that e.g. `cmake.configureArgs` supports substitution, but it seems not to be working either, e.g. having this
```
cmake.configureArgs": [
        "-DCMAKE_C_COMPILER=${buildKitVendor}",
        "-DCMAKE_CXX_COMPILER:FILEPATH=${command:cmake.buildKit}",
    ],
```
in my `settings.json`, and having this 
```
{
	"label": "CMakeConfigureTaskWorker",
	"type": "process",
	"command": "${env:SHELL}",
	"args": [
		"-c",
		{
			"value": "\"${config:cmake.cmakePath}\" --install-prefix \"${config:cmake.installPrefix}\" -S\"${workspaceFolder}\" -B\"${config:cmake.buildDirectory}\" -DCMAKE_BUILD_TYPE:STRING=${command:cmake.buildType} ${config:cmake.configureArgs} -G \"Unix Makefiles\" ", 
			"quoting": "strong"
		}
	],
```
in my tasks.json does not contain the proper value in the expanded `${config:cmake.configureArgs}` 

Where can I reference the [Configure CMake Tools settings](https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/cmake-settings.md#configure-cmake-tools-settings) variables exactly?

TIA!

p.s.: I'm using v1.15.31 on macOS, and yes, I do have the Kits scanned and set up perfectly already via [Scan for kits](https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/kits.md#scan-for-kits)

**Comment by gcampbell-msft** [comment, marked as answer]
@HofiOne I believe you are attempting to use them in situations when they are accessible, however, the locations you are using them, they need to be accessed via commands rather than just variables: https://github.com/microsoft/vscode-cmake-tools/blob/0378fdc4a409c63cfb92824f39fe3c23c9d5e0d0/docs/cmake-settings.md#command-substitution

**Reply by HofiOne** [reply]
please see my example

"-DCMAKE_CXX_COMPILER:FILEPATH=${command:cmake.buildKit}"

as I mentioned, it does not work either

**Reply by gcampbell-msft** [reply]
@HofiOne Understood, and revisiting this, you tried `${buildKit}` inside of the `cmake.configureArgs` right?

**Reply by gcampbell-msft** [reply]
Looking at your settings.json from above, what you're seeing is expected `{config:cmake.configureArgs}`, because we don't control that accessor or expansion, this is a VS Code idea that I'm not sure we can customize.

**Reply by HofiOne** [reply]
I understand, but that is why the original question was about "Where exactly these variables can be used?"
Instead of guessing and adding further attempts at various places, it would be nice to see this e.g. in the documentation.

**Reply by gcampbell-msft** [reply]
I understand, overall it's difficult to express every setting where these types of variables can be used, but I agree that it would be nice rather than having to guess and have trial and error. I'll create an issue from this and we can try to put it on our backlog.

**Reply by HofiOne** [reply]
thank you very much!
